hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/FastSelectionKeys.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright (c) 2008-2015, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.nio.tcp.nonblocking;
+
+import com.hazelcast.logging.ILogger;
+
+import java.lang.reflect.Field;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.util.AbstractSet;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static java.lang.Class.forName;
+import static java.lang.System.arraycopy;
+
+/**
+ * The FastSelectionKeys optimized the Selector so less litter is being created. The Selector uses a Set, but this creates
+ * an object for every add of a selection key. With this FastSelectionKeys an array is being used since every key is going
+ * to be inserted only once.
+ *
+ * This trick comes from Netty.
+ */
+public final class FastSelectionKeys extends AbstractSet<SelectionKey> {
+    static final int INITIAL_CAPACITY = 32;
+    static final String SELECTOR_IMPL = "sun.nio.ch.SelectorImpl";
+
+    static final String HAZELCAST_IO_FASTSELECTIONKEYS_ENABLED = "hazelcast.io.fastselectionkeys.enabled";
+
+    // Using this property the FastSelectionKeys optimization can be disabled. It is disabled by default.
+    private static final boolean ENABLED = Boolean.getBoolean(HAZELCAST_IO_FASTSELECTIONKEYS_ENABLED);
+
+    // the active SelectionKeys is the one where is being added to.
+    SelectionKeys activeKeys = new SelectionKeys();
+    // the passive SelectionKeys is one that is being read using the iterator.
+    SelectionKeys passiveKeys = new SelectionKeys();
+    private final IteratorImpl iterator = new IteratorImpl();
+
+    FastSelectionKeys() {
+    }
+
+    @Override
+    public boolean add(SelectionKey o) {
+        return activeKeys.add(o);
+    }
+
+    @Override
+    public int size() {
+        return activeKeys.size;
+    }
+
+    @Override
+    public Iterator<SelectionKey> iterator() {
+        iterator.init(flip());
+        return iterator;
+    }
+
+    private SelectionKey[] flip() {
+        SelectionKeys tmp = activeKeys;
+        activeKeys = passiveKeys;
+        passiveKeys = tmp;
+
+        activeKeys.size = 0;
+        return passiveKeys.keys;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return false;
```

Comment:
I don't think that is a wise plan. Please check the original comments;  I have implemented them in the same way as Netty and Agrona.